rpc/RPCSession.java

```diff
@@ -214,37 +211,105 @@ public String database() {
             return dbName;
         }
 
-        private Database discoverDatabaseReplicas() {
-            int attempt = 0;
-            while (attempt < clusterClient.coreClients().size()) {
+        private Grakn.Transaction transactionPrimaryReplica(Grakn.Transaction.Type type, GraknOptions options) {
+            for (Address.Cluster.Server server: clusterClient.servers()) {
+                int retry = 0;
+                while (retry < MAX_RETRY_PER_SERVER) {
+                    try {
+                        Core primaryReplicaSession = coreSessions.computeIfAbsent(
+                                database.primaryReplica().id(),
+                                key -> {
+                                    LOG.debug("Opening a session to primary replica '{}'", key);
+                                    GraknClient.Core primaryReplicaClient = clusterClient.coreClient(key.address());
+                                    return primaryReplicaClient.session(key.database(), this.type, this.options);
+                                }
+                        );
+                        LOG.debug("Opening a transaction to primary replica '{}'", database.primaryReplica().id());
+                        return primaryReplicaSession.transaction(type, options);
+                    } catch (GraknClientException e) {
+                        retry++;
+                        if (e.getErrorMessage().equals(CLUSTER_REPLICA_NOT_PRIMARY)) {
+                            LOG.debug("Unable to open a session or transaction", e);
+                            database = databaseDiscover(server);
+                        } else if (e.getErrorMessage().equals(CLUSTER_NO_PRIMARY_REPLICA_YET)) {
+                            LOG.debug("Unable to open a session or transaction", e);
+                            sleepWait();
```

Comment:
I see what you're saying there. Taking that idea, what if spin it with a different approach which is by adding more context to the name - the function is only used to wait for the primary replica to be selected. Therefore I'll rename it to `waitForPrimaryReplicaSelection()`.

With the extra context in the name I no longer have to add parameters to the function since how long it needs to wait should already be clear. This way we can avoid adding the parameters and therefore making it "a generic util wait function".

Reply:
Yeah that seems reasonable.

Reply:
Fixed in https://github.com/graknlabs/client-java/pull/243/commits/641b5769611a9dd94adfbdff22a355d8b1f94677 and https://github.com/graknlabs/client-java/pull/243/commits/5dd3a4dbf4413e1afcad8ab06f31183505f04912.